// @edgestore/shared/dist/internals/bucketBuilder.mjs
import { z } from 'zod';
import { createPathParamProxy } from './createPathParamProxy.mjs';

const createNewBuilder = (initDef, newDef)=>{
    const mergedDef = {
        ...initDef,
        ...newDef
    };
    return createBuilder({
        type: mergedDef.type
    }, mergedDef);
};
function createBuilder(opts, initDef) {
    const _def = {
        type: opts.type,
        input: z.never(),
        path: [],
        metadata: ()=>({}),
        ...initDef
    };
    return {
        $config: {
            ctx: undefined
        },
        // @ts-expect-error - I think it would be too much work to make this type correct.
        _def,
        input (input) {
            return createNewBuilder(_def, {
                input
            });
        },
        path (pathResolver) {
            // TODO: Should throw a runtime error in the following cases:
            // 1. in case of multiple keys in one object
            // 2. in case of duplicate keys
            const pathParamProxy = createPathParamProxy();
            const params = pathResolver(pathParamProxy);
            return createNewBuilder(_def, {
                path: params
            });
        },
        metadata (metadata) {
            return createNewBuilder(_def, {
                metadata
            });
        },
        accessControl (accessControl) {
            return createNewBuilder(_def, {
                accessControl: accessControl
            });
        },
        beforeUpload (beforeUpload) {
            return createNewBuilder(_def, {
                beforeUpload
            });
        },
        beforeDelete (beforeDelete) {
            return createNewBuilder(_def, {
                beforeDelete
            });
        }
    };
}
class EdgeStoreBuilder {
    context() {
        return new EdgeStoreBuilder();
    }
    create() {
        return createEdgeStoreInner()();
    }
}
function createRouterFactory() {
    return function createRouterInner(buckets) {
        return {
            $config: {
                ctx: undefined
            },
            buckets
        };
    };
}
function initBucket(type, config) {
    return createBuilder({
        type
    }, {
        bucketConfig: config
    });
}
function createEdgeStoreInner() {
    return function initEdgeStoreInner() {
        return {
            /**
       * Builder object for creating an image bucket
       */ imageBucket (config) {
                return initBucket('IMAGE', config);
            },
            /**
       * Builder object for creating a file bucket
       */ fileBucket (config) {
                return initBucket('FILE', config);
            },
            /**
       * Create a router
       */ router: createRouterFactory()
        };
    };
}
/**
 * Initialize EdgeStore - be done exactly once per backend
 */ const initEdgeStore = new EdgeStoreBuilder(); // ↓↓↓ TYPE TESTS ↓↓↓
 // type Context = {
 //   userId: string;
 //   userRole: 'admin' | 'visitor';
 // };
 // const es = initEdgeStore.context<Context>().create();
 // const imagesBucket = es.imageBucket()
 //   .input(
 //     z.object({
 //       type: z.enum(['profile', 'post']),
 //       extension: z.string().optional(),
 //     }),
 //   )
 //   .path(({ ctx, input }) => [{ author: ctx.userId }, { type: input.type }])
 //   .metadata(({ ctx, input }) => ({
 //     extension: input.extension,
 //     role: ctx.userRole,
 //   }))
 //   .beforeUpload(() => {
 //     return true;
 //   });
 // const a = es.imageBucket()
 //   .input(z.object({ type: z.string(), someMeta: z.string().optional() }))
 //   .path(({ ctx, input }) => [{ author: ctx.userId }, { type: input.type }])
 //   .metadata(({ ctx, input }) => ({
 //     role: ctx.userRole,
 //     someMeta: input.someMeta,
 //   }))
 //   .accessControl({
 //     OR: [
 //       {
 //         userId: { path: 'author' }, // this will check if the userId is the same as the author in the path parameter
 //       },
 //       {
 //         userRole: 'admin', // this is the same as { userRole: { eq: "admin" } }
 //       },
 //     ],
 //   })
 //   .beforeUpload(({ ctx, input }) => {
 //     return true;
 //   })
 //   .beforeDelete(({ ctx, file }) => {
 //     return true;
 //   });
 // const b = es.imageBucket().path(({ ctx }) => [{ author: ctx.userId }]);
 // const router = es.router({
 //   original: imagesBucket,
 //   imageBucket: a,
 //   imageBucket2: b,
 // });
 // export { router };
 // type ListFilesResponse<TBucket extends AnyRouter['buckets'][string]> = {
 //   data: {
 //     // url: string;
 //     // size: number;
 //     // uploadedAt: Date;
 //     // metadata: InferMetadataObject<TBucket>;
 //     path: InferBucketPathKeys<TBucket> extends string ? {
 //       [key: string]: string;
 //     } :{
 //       [TKey in InferBucketPathKeys<TBucket>]: string;
 //     };
 //   }[];
 //   pagination: {
 //     currentPage: number;
 //     totalPages: number;
 //     totalCount: number;
 //   };
 // };
 // type TPathKeys = 'author' | 'type';
 // type TPathKeys2 = InferBucketPathKeys<AnyBuilder>;
 // type ObjectWithKeys<TKeys extends string> = {
 //   [TKey in TKeys]: string;
 // };
 // type Test1 = ObjectWithKeys<TPathKeys>;
 // type Test2 = ObjectWithKeys<TPathKeys2>;
 // type PathKeys = InferBucketPathKeys<typeof router.buckets.imageBucket>;
 // type MetadataKeys = InferMetadataObject<typeof router.buckets.imageBucket>;
 // type MyEdgeStoreRouter = typeof router;
 // type MyAccessControl = AccessControlSchema<Context, AnyDef>;

export { initEdgeStore };
