// @edgestore/server/dist/providers/edgestore/index.js
'use strict';

var shared$1 = require('@edgestore/shared');
var shared = require('../../adapters/shared.js');
var index = require('../../core/sdk/index.js');
var EdgeStoreCredentialsError = require('../../libs/errors/EdgeStoreCredentialsError.js');

const DEFAULT_BASE_URL = 'https://files.edgestore.dev';
function EdgeStoreProvider(options) {
    const { accessKey = shared.getEnv('EDGE_STORE_ACCESS_KEY') ?? // @ts-expect-error - In Vite/Astro, the env variables are available on `import.meta`.
    undefined?.EDGE_STORE_ACCESS_KEY, secretKey = shared.getEnv('EDGE_STORE_SECRET_KEY') ?? // @ts-expect-error - In Vite/Astro, the env variables are available on `import.meta`.
    undefined?.EDGE_STORE_SECRET_KEY } = options ?? {};
    const baseUrl = shared.getEnv('EDGE_STORE_BASE_URL') ?? DEFAULT_BASE_URL;
    if (!accessKey || !secretKey) {
        throw new EdgeStoreCredentialsError();
    }
    const edgeStoreSdk = index.initEdgeStoreSdk({
        accessKey,
        secretKey
    });
    return {
        name: 'edgestore',
        init: async ({ ctx, router })=>{
            const token = await edgeStoreSdk.getToken({
                ctx,
                router
            });
            return {
                token
            };
        },
        getBaseUrl () {
            return baseUrl;
        },
        getFile: async ({ url })=>{
            const { uploadedAt, ...rest } = await edgeStoreSdk.getFile({
                url
            });
            return {
                uploadedAt: new Date(uploadedAt),
                ...rest
            };
        },
        async requestUpload ({ bucketName, bucketType, fileInfo }) {
            // multipart upload if file is bigger than a certain size
            const MULTIPART_THRESHOLD = 10 * 1024 * 1024; // 10MB
            let partSize = 5 * 1024 * 1024; // 5MB
            if (fileInfo.size > MULTIPART_THRESHOLD) {
                let totalParts = Math.ceil(fileInfo.size / partSize);
                if (totalParts > 1000) {
                    // the maximum number of parts is 1000
                    totalParts = 1000;
                    partSize = Math.ceil(fileInfo.size / totalParts);
                }
                const res = await edgeStoreSdk.requestUpload({
                    bucketName,
                    bucketType,
                    fileInfo,
                    multipart: {
                        parts: Array.from({
                            length: totalParts
                        }).map((_, index)=>index + 1)
                    }
                });
                const multipart = res.multipart ? {
                    key: res.multipart.key,
                    uploadId: res.multipart.uploadId,
                    parts: res.multipart.parts.map((part)=>({
                            partNumber: part.partNumber,
                            uploadUrl: part.signedUrl
                        })),
                    partSize,
                    totalParts
                } : undefined;
                if (multipart) {
                    return {
                        accessUrl: res.accessUrl,
                        thumbnailUrl: res.thumbnailUrl,
                        multipart
                    };
                } else if (res.signedUrl) {
                    return {
                        accessUrl: res.accessUrl,
                        uploadUrl: res.signedUrl,
                        thumbnailUrl: res.thumbnailUrl
                    };
                } else {
                    throw new shared$1.EdgeStoreError({
                        message: 'Could not get upload url',
                        code: 'SERVER_ERROR'
                    });
                }
            }
            const res = await edgeStoreSdk.requestUpload({
                bucketName,
                bucketType,
                fileInfo
            });
            if (res.signedUrl) {
                return {
                    accessUrl: res.accessUrl,
                    uploadUrl: res.signedUrl,
                    thumbnailUrl: res.thumbnailUrl
                };
            }
            throw new shared$1.EdgeStoreError({
                message: 'Could not get upload url',
                code: 'SERVER_ERROR'
            });
        },
        requestUploadParts: async ({ multipart, path })=>{
            const res = await edgeStoreSdk.requestUploadParts({
                multipart,
                key: path
            });
            return {
                multipart: {
                    uploadId: res.multipart.uploadId,
                    parts: res.multipart.parts.map((part)=>({
                            partNumber: part.partNumber,
                            uploadUrl: part.signedUrl
                        }))
                }
            };
        },
        completeMultipartUpload: async ({ uploadId, key, parts })=>{
            return await edgeStoreSdk.completeMultipartUpload({
                uploadId,
                key,
                parts
            });
        },
        confirmUpload: async ({ url })=>{
            return await edgeStoreSdk.confirmUpload({
                url
            });
        },
        deleteFile: async ({ url })=>{
            return await edgeStoreSdk.deleteFile({
                url
            });
        }
    };
}

exports.EdgeStoreProvider = EdgeStoreProvider;
