// @edgestore/server/dist/adapters/hono/index.mjs
import { EdgeStoreError, EDGE_STORE_ERROR_CODES } from '@edgestore/shared';
import Logger from '../../libs/logger.mjs';
import { matchPath } from '../../libs/utils.mjs';
import { EdgeStoreProvider } from '../../providers/edgestore/index.mjs';
import { getCookieConfig, init, requestUpload, requestUploadParts, completeMultipartUpload, confirmUpload, deleteFile } from '../shared.mjs';

// Helper to get a cookie value from Hono Context
function getCookie(c, name) {
    const cookies = c.req.header('cookie');
    if (!cookies) return undefined;
    const match = new RegExp(`${name}=([^;]+)`).exec(cookies);
    return match ? match[1] : undefined;
}
function createEdgeStoreHonoHandler(config) {
    const { provider = EdgeStoreProvider(), cookieConfig } = config;
    const log = new Logger(config.logLevel);
    globalThis._EDGE_STORE_LOGGER = log;
    log.debug('Creating EdgeStore Hono handler');
    const resolvedCookieConfig = getCookieConfig(cookieConfig);
    return async (c)=>{
        try {
            const pathname = new URL(c.req.url).pathname;
            if (matchPath(pathname, '/health')) {
                return c.text('OK');
            } else if (matchPath(pathname, '/init')) {
                let ctx = {};
                try {
                    ctx = 'createContext' in config ? await config.createContext({
                        c
                    }) : {};
                } catch (err) {
                    throw new EdgeStoreError({
                        message: 'Error creating context',
                        code: 'CREATE_CONTEXT_ERROR',
                        cause: err instanceof Error ? err : undefined
                    });
                }
                const { newCookies, token, baseUrl, providerName } = await init({
                    ctx,
                    provider,
                    router: config.router,
                    cookieConfig
                });
                // Set cookies
                if (Array.isArray(newCookies)) {
                    for (const cookie of newCookies){
                        c.header('Set-Cookie', cookie);
                    }
                } else if (newCookies) {
                    c.header('Set-Cookie', newCookies);
                }
                return c.json({
                    token,
                    baseUrl,
                    providerName
                });
            } else if (matchPath(pathname, '/request-upload')) {
                const body = await c.req.json();
                return c.json(await requestUpload({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(c, resolvedCookieConfig.ctx.name)
                }));
            } else if (matchPath(pathname, '/request-upload-parts')) {
                const body = await c.req.json();
                return c.json(await requestUploadParts({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(c, resolvedCookieConfig.ctx.name)
                }));
            } else if (matchPath(pathname, '/complete-multipart-upload')) {
                const body = await c.req.json();
                await completeMultipartUpload({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(c, resolvedCookieConfig.ctx.name)
                });
                return c.body(null, 200);
            } else if (matchPath(pathname, '/confirm-upload')) {
                const body = await c.req.json();
                return c.json(await confirmUpload({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(c, resolvedCookieConfig.ctx.name)
                }));
            } else if (matchPath(pathname, '/delete-file')) {
                const body = await c.req.json();
                return c.json(await deleteFile({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(c, resolvedCookieConfig.ctx.name)
                }));
            } else if (matchPath(pathname, '/proxy-file')) {
                const url = c.req.query('url');
                if (typeof url === 'string') {
                    const cookieHeader = c.req.header('cookie') ?? '';
                    const proxyRes = await fetch(url, {
                        headers: {
                            cookie: cookieHeader
                        }
                    });
                    const data = await proxyRes.arrayBuffer();
                    c.header('Content-Type', proxyRes.headers.get('Content-Type') ?? 'application/octet-stream');
                    return c.body(Buffer.from(data));
                } else {
                    return c.body(null, 400);
                }
            } else {
                return c.body(null, 404);
            }
        } catch (err) {
            if (err instanceof EdgeStoreError) {
                log[err.level](err.formattedMessage());
                if (err.cause) log[err.level](err.cause);
                return c.json(err.formattedJson(), EDGE_STORE_ERROR_CODES[err.code]);
            } else {
                log.error(err);
                return c.json(new EdgeStoreError({
                    message: 'Internal Server Error',
                    code: 'SERVER_ERROR'
                }).formattedJson(), 500);
            }
        }
    };
}

export { createEdgeStoreHonoHandler };
