// @edgestore/server/dist/adapters/hono/index.js
'use strict';

var shared$1 = require('@edgestore/shared');
var logger = require('../../libs/logger.js');
var utils = require('../../libs/utils.js');
var index = require('../../providers/edgestore/index.js');
var shared = require('../shared.js');

// Helper to get a cookie value from Hono Context
function getCookie(c, name) {
    const cookies = c.req.header('cookie');
    if (!cookies) return undefined;
    const match = new RegExp(`${name}=([^;]+)`).exec(cookies);
    return match ? match[1] : undefined;
}
function createEdgeStoreHonoHandler(config) {
    const { provider = index.EdgeStoreProvider(), cookieConfig } = config;
    const log = new logger(config.logLevel);
    globalThis._EDGE_STORE_LOGGER = log;
    log.debug('Creating EdgeStore Hono handler');
    const resolvedCookieConfig = shared.getCookieConfig(cookieConfig);
    return async (c)=>{
        try {
            const pathname = new URL(c.req.url).pathname;
            if (utils.matchPath(pathname, '/health')) {
                return c.text('OK');
            } else if (utils.matchPath(pathname, '/init')) {
                let ctx = {};
                try {
                    ctx = 'createContext' in config ? await config.createContext({
                        c
                    }) : {};
                } catch (err) {
                    throw new shared$1.EdgeStoreError({
                        message: 'Error creating context',
                        code: 'CREATE_CONTEXT_ERROR',
                        cause: err instanceof Error ? err : undefined
                    });
                }
                const { newCookies, token, baseUrl, providerName } = await shared.init({
                    ctx,
                    provider,
                    router: config.router,
                    cookieConfig
                });
                // Set cookies
                if (Array.isArray(newCookies)) {
                    for (const cookie of newCookies){
                        c.header('Set-Cookie', cookie);
                    }
                } else if (newCookies) {
                    c.header('Set-Cookie', newCookies);
                }
                return c.json({
                    token,
                    baseUrl,
                    providerName
                });
            } else if (utils.matchPath(pathname, '/request-upload')) {
                const body = await c.req.json();
                return c.json(await shared.requestUpload({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(c, resolvedCookieConfig.ctx.name)
                }));
            } else if (utils.matchPath(pathname, '/request-upload-parts')) {
                const body = await c.req.json();
                return c.json(await shared.requestUploadParts({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(c, resolvedCookieConfig.ctx.name)
                }));
            } else if (utils.matchPath(pathname, '/complete-multipart-upload')) {
                const body = await c.req.json();
                await shared.completeMultipartUpload({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(c, resolvedCookieConfig.ctx.name)
                });
                return c.body(null, 200);
            } else if (utils.matchPath(pathname, '/confirm-upload')) {
                const body = await c.req.json();
                return c.json(await shared.confirmUpload({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(c, resolvedCookieConfig.ctx.name)
                }));
            } else if (utils.matchPath(pathname, '/delete-file')) {
                const body = await c.req.json();
                return c.json(await shared.deleteFile({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(c, resolvedCookieConfig.ctx.name)
                }));
            } else if (utils.matchPath(pathname, '/proxy-file')) {
                const url = c.req.query('url');
                if (typeof url === 'string') {
                    const cookieHeader = c.req.header('cookie') ?? '';
                    const proxyRes = await fetch(url, {
                        headers: {
                            cookie: cookieHeader
                        }
                    });
                    const data = await proxyRes.arrayBuffer();
                    c.header('Content-Type', proxyRes.headers.get('Content-Type') ?? 'application/octet-stream');
                    return c.body(Buffer.from(data));
                } else {
                    return c.body(null, 400);
                }
            } else {
                return c.body(null, 404);
            }
        } catch (err) {
            if (err instanceof shared$1.EdgeStoreError) {
                log[err.level](err.formattedMessage());
                if (err.cause) log[err.level](err.cause);
                return c.json(err.formattedJson(), shared$1.EDGE_STORE_ERROR_CODES[err.code]);
            } else {
                log.error(err);
                return c.json(new shared$1.EdgeStoreError({
                    message: 'Internal Server Error',
                    code: 'SERVER_ERROR'
                }).formattedJson(), 500);
            }
        }
    };
}

exports.createEdgeStoreHonoHandler = createEdgeStoreHonoHandler;
