// @edgestore/server/dist/adapters/astro/index.js
'use strict';

var shared$1 = require('@edgestore/shared');
var logger = require('../../libs/logger.js');
var utils = require('../../libs/utils.js');
var index = require('../../providers/edgestore/index.js');
var shared = require('../shared.js');

// Helper to safely get cookies from Astro request
function getCookie(request, name) {
    const cookieHeader = request.headers.get('cookie');
    if (!cookieHeader) return undefined;
    const cookies = cookieHeader.split(';').reduce((acc, cookie)=>{
        const [key, value] = cookie.trim().split('=');
        if (key && value) acc[key] = value;
        return acc;
    }, {});
    return cookies[name];
}
function createEdgeStoreAstroHandler(config) {
    const { provider = index.EdgeStoreProvider(), cookieConfig } = config;
    const log = new logger(config.logLevel);
    globalThis._EDGE_STORE_LOGGER = log;
    log.debug('Creating EdgeStore Astro handler');
    const resolvedCookieConfig = shared.getCookieConfig(cookieConfig);
    return async (context)=>{
        try {
            const { request } = context;
            const url = new URL(request.url);
            if (utils.matchPath(url.pathname, 'health')) {
                return new Response('OK');
            } else if (utils.matchPath(url.pathname, 'init')) {
                let ctx = {};
                try {
                    ctx = 'createContext' in config ? await config.createContext(context) : {};
                } catch (err) {
                    throw new shared$1.EdgeStoreError({
                        message: 'Error creating context',
                        code: 'CREATE_CONTEXT_ERROR',
                        cause: err instanceof Error ? err : undefined
                    });
                }
                const { newCookies, token, baseUrl, providerName } = await shared.init({
                    ctx,
                    provider,
                    router: config.router,
                    cookieConfig
                });
                const headers = new Headers();
                headers.set('Content-Type', 'application/json');
                // Set cookies
                if (Array.isArray(newCookies)) {
                    for (const cookie of newCookies){
                        headers.append('Set-Cookie', cookie);
                    }
                } else if (newCookies) {
                    headers.append('Set-Cookie', newCookies);
                }
                return new Response(JSON.stringify({
                    token,
                    baseUrl,
                    providerName
                }), {
                    headers
                });
            } else if (utils.matchPath(url.pathname, 'request-upload')) {
                const body = await request.json();
                const result = await shared.requestUpload({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(request, resolvedCookieConfig.ctx.name)
                });
                return new Response(JSON.stringify(result), {
                    headers: {
                        'Content-Type': 'application/json'
                    }
                });
            } else if (utils.matchPath(url.pathname, 'request-upload-parts')) {
                const body = await request.json();
                const result = await shared.requestUploadParts({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(request, resolvedCookieConfig.ctx.name)
                });
                return new Response(JSON.stringify(result), {
                    headers: {
                        'Content-Type': 'application/json'
                    }
                });
            } else if (utils.matchPath(url.pathname, 'complete-multipart-upload')) {
                const body = await request.json();
                await shared.completeMultipartUpload({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(request, resolvedCookieConfig.ctx.name)
                });
                return new Response(null, {
                    status: 200
                });
            } else if (utils.matchPath(url.pathname, 'confirm-upload')) {
                const body = await request.json();
                const result = await shared.confirmUpload({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(request, resolvedCookieConfig.ctx.name)
                });
                return new Response(JSON.stringify(result), {
                    headers: {
                        'Content-Type': 'application/json'
                    }
                });
            } else if (utils.matchPath(url.pathname, 'delete-file')) {
                const body = await request.json();
                const result = await shared.deleteFile({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(request, resolvedCookieConfig.ctx.name)
                });
                return new Response(JSON.stringify(result), {
                    headers: {
                        'Content-Type': 'application/json'
                    }
                });
            } else if (utils.matchPath(url.pathname, 'proxy-file')) {
                const url = new URL(request.url).searchParams.get('url');
                if (typeof url === 'string') {
                    const cookieHeader = request.headers.get('cookie') ?? '';
                    const proxyRes = await fetch(url, {
                        headers: {
                            cookie: cookieHeader
                        }
                    });
                    const data = await proxyRes.arrayBuffer();
                    const headers = new Headers();
                    headers.set('Content-Type', proxyRes.headers.get('Content-Type') ?? 'application/octet-stream');
                    return new Response(data, {
                        headers
                    });
                } else {
                    return new Response(null, {
                        status: 400
                    });
                }
            } else {
                return new Response(null, {
                    status: 404
                });
            }
        } catch (err) {
            if (err instanceof shared$1.EdgeStoreError) {
                log[err.level](err.formattedMessage());
                if (err.cause) log[err.level](err.cause);
                return new Response(JSON.stringify(err.formattedJson()), {
                    status: shared$1.EDGE_STORE_ERROR_CODES[err.code],
                    headers: {
                        'Content-Type': 'application/json'
                    }
                });
            } else {
                log.error(err);
                return new Response(JSON.stringify(new shared$1.EdgeStoreError({
                    message: 'Internal Server Error',
                    code: 'SERVER_ERROR'
                }).formattedJson()), {
                    status: 500,
                    headers: {
                        'Content-Type': 'application/json'
                    }
                });
            }
        }
    };
}

exports.createEdgeStoreAstroHandler = createEdgeStoreAstroHandler;
