// @edgestore/server/dist/adapters/remix/index.js
'use strict';

var shared$1 = require('@edgestore/shared');
var logger = require('../../libs/logger.js');
var utils = require('../../libs/utils.js');
var index = require('../../providers/edgestore/index.js');
var shared = require('../shared.js');

// Helper to safely get cookies from Remix request
function getCookie(req, name) {
    const cookieHeader = req.headers.get('cookie');
    if (!cookieHeader) return undefined;
    const cookies = cookieHeader.split(';').reduce((acc, cookie)=>{
        const [key, value] = cookie.trim().split('=');
        if (key && value) acc[key] = value;
        return acc;
    }, {});
    return cookies[name];
}
function createEdgeStoreRemixHandler(config) {
    const { provider = index.EdgeStoreProvider(), cookieConfig } = config;
    const log = new logger(config.logLevel);
    globalThis._EDGE_STORE_LOGGER = log;
    log.debug('Creating EdgeStore Remix handler');
    const resolvedCookieConfig = shared.getCookieConfig(cookieConfig);
    return async ({ request: req })=>{
        try {
            const url = new URL(req.url);
            const pathname = url.pathname;
            if (utils.matchPath(pathname, '/health')) {
                return new Response('OK');
            } else if (utils.matchPath(pathname, '/init')) {
                let ctx = {};
                try {
                    ctx = 'createContext' in config ? await config.createContext({
                        req
                    }) : {};
                } catch (err) {
                    throw new shared$1.EdgeStoreError({
                        message: 'Error creating context',
                        code: 'CREATE_CONTEXT_ERROR',
                        cause: err instanceof Error ? err : undefined
                    });
                }
                const { newCookies, token, baseUrl, providerName } = await shared.init({
                    ctx,
                    provider,
                    router: config.router,
                    cookieConfig
                });
                // Create response with cookies and token
                const responseHeaders = new Headers();
                if (Array.isArray(newCookies)) {
                    for (const cookie of newCookies){
                        responseHeaders.append('Set-Cookie', cookie);
                    }
                } else if (newCookies) {
                    responseHeaders.append('Set-Cookie', newCookies);
                }
                return new Response(JSON.stringify({
                    token,
                    baseUrl,
                    providerName
                }), {
                    headers: responseHeaders,
                    status: 200
                });
            } else if (utils.matchPath(pathname, '/request-upload')) {
                const body = await req.json();
                return Response.json(await shared.requestUpload({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                }));
            } else if (utils.matchPath(pathname, '/request-upload-parts')) {
                const body = await req.json();
                return Response.json(await shared.requestUploadParts({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                }));
            } else if (utils.matchPath(pathname, '/complete-multipart-upload')) {
                const body = await req.json();
                await shared.completeMultipartUpload({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                });
                return new Response(null, {
                    status: 200
                });
            } else if (utils.matchPath(pathname, '/confirm-upload')) {
                const body = await req.json();
                return Response.json(await shared.confirmUpload({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                }));
            } else if (utils.matchPath(pathname, '/delete-file')) {
                const body = await req.json();
                return Response.json(await shared.deleteFile({
                    provider,
                    router: config.router,
                    body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                }));
            } else if (utils.matchPath(pathname, '/proxy-file')) {
                const url = new URL(req.url).searchParams.get('url');
                if (typeof url === 'string') {
                    const proxyRes = await fetch(url, {
                        headers: {
                            cookie: req.headers.get('cookie') ?? ''
                        }
                    });
                    const data = await proxyRes.arrayBuffer();
                    const headers = new Headers();
                    headers.set('Content-Type', proxyRes.headers.get('Content-Type') ?? 'application/octet-stream');
                    return new Response(data, {
                        headers
                    });
                } else {
                    return new Response(null, {
                        status: 400
                    });
                }
            } else {
                return new Response(null, {
                    status: 404
                });
            }
        } catch (err) {
            if (err instanceof shared$1.EdgeStoreError) {
                log[err.level](err.formattedMessage());
                if (err.cause) log[err.level](err.cause);
                return Response.json(err.formattedJson(), {
                    status: shared$1.EDGE_STORE_ERROR_CODES[err.code]
                });
            } else {
                log.error(err);
                return Response.json(new shared$1.EdgeStoreError({
                    message: 'Internal Server Error',
                    code: 'SERVER_ERROR'
                }).formattedJson(), {
                    status: 500
                });
            }
        }
    };
}

exports.createEdgeStoreRemixHandler = createEdgeStoreRemixHandler;
