// @edgestore/server/dist/adapters/next/app/index.js
'use strict';

var shared$1 = require('@edgestore/shared');
var logger = require('../../../libs/logger.js');
var utils = require('../../../libs/utils.js');
var index = require('../../../providers/edgestore/index.js');
var shared = require('../../shared.js');

function createEdgeStoreNextHandler(config) {
    const { provider = index.EdgeStoreProvider(), cookieConfig } = config;
    const log = new logger(config.logLevel);
    globalThis._EDGE_STORE_LOGGER = log;
    log.debug('Creating EdgeStore Next handler (app adapter)');
    const resolvedCookieConfig = shared.getCookieConfig(cookieConfig);
    return async (req)=>{
        try {
            if (!('nextUrl' in req)) throw new shared$1.EdgeStoreError({
                message: 'Error running the app adapter. Make sure you are importing the correct adapter in your router configuration',
                code: 'SERVER_ERROR'
            });
            const pathname = req.nextUrl.pathname;
            if (utils.matchPath(pathname, '/health')) {
                return new Response('OK', {
                    status: 200
                });
            } else if (utils.matchPath(pathname, '/init')) {
                let ctx = {};
                try {
                    ctx = 'createContext' in config ? await config.createContext({
                        req
                    }) : {};
                } catch (err) {
                    throw new shared$1.EdgeStoreError({
                        message: 'Error creating context',
                        code: 'CREATE_CONTEXT_ERROR',
                        cause: err instanceof Error ? err : undefined
                    });
                }
                const { newCookies, token, baseUrl, providerName } = await shared.init({
                    ctx,
                    provider,
                    router: config.router,
                    cookieConfig
                });
                const res = new Response(JSON.stringify({
                    token,
                    baseUrl,
                    providerName
                }), {
                    status: 200,
                    headers: {
                        'Content-Type': 'application/json'
                    }
                });
                for (const cookie of newCookies){
                    res.headers.append('Set-Cookie', cookie);
                }
                return res;
            } else if (utils.matchPath(pathname, '/request-upload')) {
                const res = await shared.requestUpload({
                    provider,
                    router: config.router,
                    body: await req.json(),
                    ctxToken: req.cookies.get(resolvedCookieConfig.ctx.name)?.value
                });
                return new Response(JSON.stringify(res), {
                    status: 200,
                    headers: {
                        'Content-Type': 'application/json'
                    }
                });
            } else if (utils.matchPath(pathname, '/request-upload-parts')) {
                const res = await shared.requestUploadParts({
                    provider,
                    router: config.router,
                    body: await req.json(),
                    ctxToken: req.cookies.get(resolvedCookieConfig.ctx.name)?.value
                });
                return new Response(JSON.stringify(res), {
                    status: 200,
                    headers: {
                        'Content-Type': 'application/json'
                    }
                });
            } else if (utils.matchPath(pathname, '/complete-multipart-upload')) {
                await shared.completeMultipartUpload({
                    provider,
                    router: config.router,
                    body: await req.json(),
                    ctxToken: req.cookies.get(resolvedCookieConfig.ctx.name)?.value
                });
                return new Response(null, {
                    status: 200
                });
            } else if (utils.matchPath(pathname, '/confirm-upload')) {
                const res = await shared.confirmUpload({
                    provider,
                    router: config.router,
                    body: await req.json(),
                    ctxToken: req.cookies.get(resolvedCookieConfig.ctx.name)?.value
                });
                return new Response(JSON.stringify(res), {
                    status: 200,
                    headers: {
                        'Content-Type': 'application/json'
                    }
                });
            } else if (utils.matchPath(pathname, '/delete-file')) {
                const res = await shared.deleteFile({
                    provider,
                    router: config.router,
                    body: await req.json(),
                    ctxToken: req.cookies.get(resolvedCookieConfig.ctx.name)?.value
                });
                return new Response(JSON.stringify(res), {
                    status: 200,
                    headers: {
                        'Content-Type': 'application/json'
                    }
                });
            } else if (utils.matchPath(pathname, '/proxy-file')) {
                const url = req.nextUrl.searchParams.get('url');
                if (typeof url === 'string') {
                    const proxyRes = await fetch(url, {
                        headers: {
                            cookie: req.cookies.toString() ?? ''
                        }
                    });
                    const data = await proxyRes.arrayBuffer();
                    return new Response(data, {
                        status: proxyRes.status,
                        headers: {
                            'Content-Type': proxyRes.headers.get('Content-Type') ?? 'application/octet-stream'
                        }
                    });
                } else {
                    return new Response(null, {
                        status: 400
                    });
                }
            } else {
                return new Response(null, {
                    status: 404
                });
            }
        } catch (err) {
            if (err instanceof shared$1.EdgeStoreError) {
                log[err.level](err.formattedMessage());
                if (err.cause) log[err.level](err.cause);
                return new Response(JSON.stringify(err.formattedJson()), {
                    status: shared$1.EDGE_STORE_ERROR_CODES[err.code],
                    headers: {
                        'Content-Type': 'application/json'
                    }
                });
            }
            log.error(err);
            return new Response(JSON.stringify(new shared$1.EdgeStoreError({
                message: 'Internal server error',
                code: 'SERVER_ERROR'
            }).formattedJson()), {
                status: 500,
                headers: {
                    'Content-Type': 'application/json'
                }
            });
        }
    };
}

exports.createEdgeStoreNextHandler = createEdgeStoreNextHandler;
