// @edgestore/server/dist/adapters/fastify/index.js
'use strict';

var shared$1 = require('@edgestore/shared');
var logger = require('../../libs/logger.js');
var utils = require('../../libs/utils.js');
var index = require('../../providers/edgestore/index.js');
var shared = require('../shared.js');

// Helper to safely get cookies from Fastify request
function getCookie(req, name) {
    // Check if cookies plugin is available
    if ('cookies' in req) {
        // Type assertion for TypeScript
        return req.cookies[name];
    }
    // Fallback to parsing cookie header
    const cookieHeader = req.headers.cookie;
    if (!cookieHeader) return undefined;
    const cookies = cookieHeader.split(';').reduce((acc, cookie)=>{
        const [key, value] = cookie.trim().split('=');
        if (key && value) acc[key] = value;
        return acc;
    }, {});
    return cookies[name];
}
function createEdgeStoreFastifyHandler(config) {
    const { provider = index.EdgeStoreProvider(), cookieConfig } = config;
    const log = new logger(config.logLevel);
    globalThis._EDGE_STORE_LOGGER = log;
    log.debug('Creating EdgeStore Fastify handler');
    const resolvedCookieConfig = shared.getCookieConfig(cookieConfig);
    return async (req, reply)=>{
        try {
            // Get the URL from the request - simplified approach
            const pathname = req.url;
            if (utils.matchPath(pathname, '/health')) {
                return reply.send('OK');
            } else if (utils.matchPath(pathname, '/init')) {
                let ctx = {};
                try {
                    ctx = 'createContext' in config ? await config.createContext({
                        req,
                        reply
                    }) : {};
                } catch (err) {
                    throw new shared$1.EdgeStoreError({
                        message: 'Error creating context',
                        code: 'CREATE_CONTEXT_ERROR',
                        cause: err instanceof Error ? err : undefined
                    });
                }
                const { newCookies, token, baseUrl, providerName } = await shared.init({
                    ctx,
                    provider,
                    router: config.router,
                    cookieConfig
                });
                // Set cookies more efficiently - handling them using void operator
                // to explicitly mark these synchronous calls as intentionally not awaited
                if (Array.isArray(newCookies)) {
                    // If it's an array of cookies, set them all
                    for (const cookie of newCookies){
                        void reply.header('Set-Cookie', cookie);
                    }
                } else if (newCookies) {
                    // If it's a single cookie string
                    void reply.header('Set-Cookie', newCookies);
                }
                return reply.send({
                    token,
                    baseUrl,
                    providerName
                });
            } else if (utils.matchPath(pathname, '/request-upload')) {
                return reply.send(await shared.requestUpload({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                }));
            } else if (utils.matchPath(pathname, '/request-upload-parts')) {
                return reply.send(await shared.requestUploadParts({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                }));
            } else if (utils.matchPath(pathname, '/complete-multipart-upload')) {
                await shared.completeMultipartUpload({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                });
                return reply.status(200).send();
            } else if (utils.matchPath(pathname, '/confirm-upload')) {
                return reply.send(await shared.confirmUpload({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                }));
            } else if (utils.matchPath(pathname, '/delete-file')) {
                return reply.send(await shared.deleteFile({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                }));
            } else if (utils.matchPath(pathname, '/proxy-file')) {
                const url = req.query ? req.query.url : undefined;
                if (typeof url === 'string') {
                    const cookieHeader = req.headers.cookie ?? '';
                    const proxyRes = await fetch(url, {
                        headers: {
                            cookie: cookieHeader
                        }
                    });
                    const data = await proxyRes.arrayBuffer();
                    void reply.header('Content-Type', proxyRes.headers.get('Content-Type') ?? 'application/octet-stream');
                    return reply.send(Buffer.from(data));
                } else {
                    return reply.status(400).send();
                }
            } else {
                return reply.status(404).send();
            }
        } catch (err) {
            if (err instanceof shared$1.EdgeStoreError) {
                log[err.level](err.formattedMessage());
                if (err.cause) log[err.level](err.cause);
                return reply.status(shared$1.EDGE_STORE_ERROR_CODES[err.code]).send(err.formattedJson());
            } else {
                log.error(err);
                return reply.status(500).send(new shared$1.EdgeStoreError({
                    message: 'Internal Server Error',
                    code: 'SERVER_ERROR'
                }).formattedJson());
            }
        }
    };
}

exports.createEdgeStoreFastifyHandler = createEdgeStoreFastifyHandler;
