// @edgestore/server/dist/adapters/fastify/index.mjs
import { EdgeStoreError, EDGE_STORE_ERROR_CODES } from '@edgestore/shared';
import Logger from '../../libs/logger.mjs';
import { matchPath } from '../../libs/utils.mjs';
import { EdgeStoreProvider } from '../../providers/edgestore/index.mjs';
import { getCookieConfig, init, requestUpload, requestUploadParts, completeMultipartUpload, confirmUpload, deleteFile } from '../shared.mjs';

// Helper to safely get cookies from Fastify request
function getCookie(req, name) {
    // Check if cookies plugin is available
    if ('cookies' in req) {
        // Type assertion for TypeScript
        return req.cookies[name];
    }
    // Fallback to parsing cookie header
    const cookieHeader = req.headers.cookie;
    if (!cookieHeader) return undefined;
    const cookies = cookieHeader.split(';').reduce((acc, cookie)=>{
        const [key, value] = cookie.trim().split('=');
        if (key && value) acc[key] = value;
        return acc;
    }, {});
    return cookies[name];
}
function createEdgeStoreFastifyHandler(config) {
    const { provider = EdgeStoreProvider(), cookieConfig } = config;
    const log = new Logger(config.logLevel);
    globalThis._EDGE_STORE_LOGGER = log;
    log.debug('Creating EdgeStore Fastify handler');
    const resolvedCookieConfig = getCookieConfig(cookieConfig);
    return async (req, reply)=>{
        try {
            // Get the URL from the request - simplified approach
            const pathname = req.url;
            if (matchPath(pathname, '/health')) {
                return reply.send('OK');
            } else if (matchPath(pathname, '/init')) {
                let ctx = {};
                try {
                    ctx = 'createContext' in config ? await config.createContext({
                        req,
                        reply
                    }) : {};
                } catch (err) {
                    throw new EdgeStoreError({
                        message: 'Error creating context',
                        code: 'CREATE_CONTEXT_ERROR',
                        cause: err instanceof Error ? err : undefined
                    });
                }
                const { newCookies, token, baseUrl, providerName } = await init({
                    ctx,
                    provider,
                    router: config.router,
                    cookieConfig
                });
                // Set cookies more efficiently - handling them using void operator
                // to explicitly mark these synchronous calls as intentionally not awaited
                if (Array.isArray(newCookies)) {
                    // If it's an array of cookies, set them all
                    for (const cookie of newCookies){
                        void reply.header('Set-Cookie', cookie);
                    }
                } else if (newCookies) {
                    // If it's a single cookie string
                    void reply.header('Set-Cookie', newCookies);
                }
                return reply.send({
                    token,
                    baseUrl,
                    providerName
                });
            } else if (matchPath(pathname, '/request-upload')) {
                return reply.send(await requestUpload({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                }));
            } else if (matchPath(pathname, '/request-upload-parts')) {
                return reply.send(await requestUploadParts({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                }));
            } else if (matchPath(pathname, '/complete-multipart-upload')) {
                await completeMultipartUpload({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                });
                return reply.status(200).send();
            } else if (matchPath(pathname, '/confirm-upload')) {
                return reply.send(await confirmUpload({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                }));
            } else if (matchPath(pathname, '/delete-file')) {
                return reply.send(await deleteFile({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: getCookie(req, resolvedCookieConfig.ctx.name)
                }));
            } else if (matchPath(pathname, '/proxy-file')) {
                const url = req.query ? req.query.url : undefined;
                if (typeof url === 'string') {
                    const cookieHeader = req.headers.cookie ?? '';
                    const proxyRes = await fetch(url, {
                        headers: {
                            cookie: cookieHeader
                        }
                    });
                    const data = await proxyRes.arrayBuffer();
                    void reply.header('Content-Type', proxyRes.headers.get('Content-Type') ?? 'application/octet-stream');
                    return reply.send(Buffer.from(data));
                } else {
                    return reply.status(400).send();
                }
            } else {
                return reply.status(404).send();
            }
        } catch (err) {
            if (err instanceof EdgeStoreError) {
                log[err.level](err.formattedMessage());
                if (err.cause) log[err.level](err.cause);
                return reply.status(EDGE_STORE_ERROR_CODES[err.code]).send(err.formattedJson());
            } else {
                log.error(err);
                return reply.status(500).send(new EdgeStoreError({
                    message: 'Internal Server Error',
                    code: 'SERVER_ERROR'
                }).formattedJson());
            }
        }
    };
}

export { createEdgeStoreFastifyHandler };
