// @edgestore/server/dist/adapters/next/pages/index.js
'use strict';

var shared$1 = require('@edgestore/shared');
var logger = require('../../../libs/logger.js');
var utils = require('../../../libs/utils.js');
var index = require('../../../providers/edgestore/index.js');
var shared = require('../../shared.js');

function createEdgeStoreNextHandler(config) {
    const { provider = index.EdgeStoreProvider(), cookieConfig } = config;
    const log = new logger(config.logLevel);
    globalThis._EDGE_STORE_LOGGER = log;
    log.debug('Creating EdgeStore Next handler (pages adapter)');
    const resolvedCookieConfig = shared.getCookieConfig(cookieConfig);
    return async (req, res)=>{
        try {
            if (!('json' in res)) throw new shared$1.EdgeStoreError({
                message: 'Error running the pages adapter. Make sure you are importing the correct adapter in your router configuration',
                code: 'SERVER_ERROR'
            });
            const pathname = req.url ?? '';
            if (utils.matchPath(pathname, '/health')) {
                res.send('OK');
            } else if (utils.matchPath(pathname, '/init')) {
                let ctx = {};
                try {
                    ctx = 'createContext' in config ? await config.createContext({
                        req,
                        res
                    }) : {};
                } catch (err) {
                    throw new shared$1.EdgeStoreError({
                        message: 'Error creating context',
                        code: 'CREATE_CONTEXT_ERROR',
                        cause: err instanceof Error ? err : undefined
                    });
                }
                const { newCookies, token, baseUrl, providerName } = await shared.init({
                    ctx,
                    provider,
                    router: config.router,
                    cookieConfig
                });
                res.setHeader('Set-Cookie', newCookies);
                res.json({
                    token,
                    baseUrl,
                    providerName
                });
            } else if (utils.matchPath(pathname, '/request-upload')) {
                res.json(await shared.requestUpload({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: req.cookies[resolvedCookieConfig.ctx.name]
                }));
            } else if (utils.matchPath(pathname, '/request-upload-parts')) {
                res.json(await shared.requestUploadParts({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: req.cookies[resolvedCookieConfig.ctx.name]
                }));
            } else if (utils.matchPath(pathname, '/complete-multipart-upload')) {
                await shared.completeMultipartUpload({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: req.cookies[resolvedCookieConfig.ctx.name]
                });
                res.status(200).end();
            } else if (utils.matchPath(pathname, '/confirm-upload')) {
                res.json(await shared.confirmUpload({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: req.cookies[resolvedCookieConfig.ctx.name]
                }));
            } else if (utils.matchPath(pathname, '/delete-file')) {
                res.json(await shared.deleteFile({
                    provider,
                    router: config.router,
                    body: req.body,
                    ctxToken: req.cookies[resolvedCookieConfig.ctx.name]
                }));
            } else if (utils.matchPath(pathname, '/proxy-file')) {
                const { url } = req.query;
                if (typeof url === 'string') {
                    const proxyRes = await fetch(url, {
                        headers: {
                            cookie: req.headers.cookie ?? ''
                        }
                    });
                    const data = await proxyRes.arrayBuffer();
                    res.setHeader('Content-Type', proxyRes.headers.get('Content-Type') ?? 'application/octet-stream');
                    res.end(Buffer.from(data));
                } else {
                    res.status(400).end();
                }
            } else {
                res.status(404).end();
            }
        } catch (err) {
            if (err instanceof shared$1.EdgeStoreError) {
                log[err.level](err.formattedMessage());
                if (err.cause) log[err.level](err.cause);
                res.status(shared$1.EDGE_STORE_ERROR_CODES[err.code]).json(err.formattedJson());
            } else {
                log.error(err);
                res.status(500).send(new shared$1.EdgeStoreError({
                    message: 'Internal Server Error',
                    code: 'SERVER_ERROR'
                }).formattedJson());
            }
        }
    };
}

exports.createEdgeStoreNextHandler = createEdgeStoreNextHandler;
