// @edgestore/shared/dist/errors/EdgeStoreError.js
'use strict';

function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
/* eslint-disable @typescript-eslint/no-non-null-assertion */ const EDGE_STORE_ERROR_CODES = {
    BAD_REQUEST: 400,
    FILE_TOO_LARGE: 400,
    MIME_TYPE_NOT_ALLOWED: 400,
    UNAUTHORIZED: 401,
    UPLOAD_NOT_ALLOWED: 403,
    DELETE_NOT_ALLOWED: 403,
    CREATE_CONTEXT_ERROR: 500,
    SERVER_ERROR: 500
};
class EdgeStoreError extends Error {
    formattedMessage() {
        return `${this.message}${this.details ? `\n    Details: ${JSON.stringify(this.details)}` : ''}${this.cause ? `\n    Caused by: ${this.cause.message}` : ''}`;
    }
    formattedJson() {
        return {
            message: this.code === 'SERVER_ERROR' ? 'Internal server error' : this.message,
            code: this.code,
            details: this.details
        };
    }
    constructor(opts){
        super(opts.message);
        _define_property(this, "cause", void 0);
        _define_property(this, "code", void 0);
        _define_property(this, "level", void 0);
        _define_property(this, "details", void 0);
        this.name = 'EdgeStoreError';
        this.code = opts.code;
        this.cause = opts.cause;
        this.level = EDGE_STORE_ERROR_CODES[opts.code] >= 500 ? 'error' : 'warn';
        this.details = 'details' in opts ? opts.details : undefined;
    }
}

exports.EDGE_STORE_ERROR_CODES = EDGE_STORE_ERROR_CODES;
exports.EdgeStoreError = EdgeStoreError;
